build.sh

```diff
@@ -6,33 +6,36 @@ else
   SED=sed
 fi
 
-stable() {
-  rustup toolchain add stable >&2
-  rustup run stable rustc --version | egrep -o '1\.[0-9]+\.[0-9]+'
+rustc_crawl() {
+  ptrn='\[pkg\.rust\]\nversion = "\K[^"]+'
+  relday=$(date -I)  # Today's date
+  # Rust stable updates happen every 6 weeks, starting at 2015-12-10.
+  # This would be a very clean method of extracting versions, were it not for patches.
+  # Rust 1.72.1 was released 21 days after 1.72.0 for example, but 1.73.0 was still
+  # on schedule. This makes last modified time of the stable file unreliable.
+  if [[ "$1" = 'stable' ]] ; then
+      rustup run stable rustc --version | grep -Eo '[0-9]+\.[0-9]+\.[0-9]+'
+      return 0
+  fi
+
+  # nightly releases should be found on the first check here, making it equivalent to an if,
+  # but beta releases are a pain in the behind to track down, so we probe a few days for those.
+  # On the off chance this ever runs before nightly builds, nightly falls back to yesterday's.
+  while [[ $(curl -s "https://static.rust-lang.org/?prefix=dist/$relday" | grep -c "$1") == '0' ]] ; do
+    relday=$(date -I -d "$relday -1 day")
+  done
```

Comment:
Uff, that means we poke around for quite some requests to find beta?
How does rustup do it?

Reply:
Turns out this is exactly what they do for nightly, potentially back for **21** days even! And they do that for both no new nightly as well as partial toolchain upgrades if not all components have received an update. For example, if a nightly install contains `clippy` and `clippy` hasn't had new builds in 11 days, rustup will happily do 12 requests to find a compatible version, or it concludes you have the latest.

As for beta, there seems to be some endpoint, but attempting to reconstruct that manually is giving me invalid keys and 404 pages. Will look at it again tomorrow unless you don't mind this approach

Reply:
If rustup does it then I mind much less. Can you then include a comment or link or something that makes it clear that we're doing just the same?

Reply:
Fixed and brought down to a single request that shouldn't fail unless rust-lang/rust does. But at that point, they _should_ still be serving yesterday's nightly considering no new builds can be produced